cql3/restrictions/analyzed_where.hh

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: (AGPL-3.0-or-later and Apache-2.0)
+ */
+
+#pragma once
+
+#include "index_table_query_restrictions.hh"
+#include "single_table_query_restrictions.hh"
+#include "statement_restrictions.hh"
+
+namespace cql3 {
+namespace restrictions {
+
+// This class analyzes the WHERE clause and generates a query plan.
+class analyzed_where_clause {
+    // The query is a single_table_query if possible without filtering, otherwise it might be an indexed query.
+    std::optional<std::variant<single_table_query_restrictions, index_table_query_restrictions>> query_restrictions;
```

Comment:
I think there should be more here - full scan should have a different analysis than single partition. We can adjust it later if needed.